sycl/include/sycl/reduction.hpp

```diff
@@ -666,15 +667,18 @@ class reduction_impl_algo : public reduction_impl_common<T, BinaryOperation> {
 
   // On discrete (vs. integrated) GPUs it's faster to initialize memory with an
   // extra kernel than copy it from the host.
-  template <typename Name> auto getGroupsCounterAccDiscrete(handler &CGH) {
-    auto &Buf = getTempBuffer<int>(1, CGH);
-    std::shared_ptr<detail::queue_impl> QueueCopy = CGH.MQueue;
-    auto Event = CGH.withAuxHandler(QueueCopy, [&](handler &InitHandler) {
-      auto Acc = accessor{Buf, InitHandler, sycl::write_only, sycl::no_init};
-      InitHandler.single_task<Name>([=]() { Acc[0] = 0; });
-    });
+  auto getGroupsCounterAccDiscrete(handler &CGH) {
+    queue q = createSyclObjFromImpl<queue>(CGH.MQueue);
+    device Dev = q.get_device();
+    auto Deleter = [=](auto *Ptr) { free(Ptr, q); };
+
+    std::shared_ptr<int> Counter(malloc_device<int>(1, q), Deleter);
```

Comment:
Here it is assumed that 'device' allocation is always supported.
If there are any real chance that it is not supported somewhere, then may need to have a copy of that older code.